src/xcode/ENA/ENA/Source/Services/TicketValidation/TicketValidation.swift

```diff
@@ -24,61 +24,167 @@ final class TicketValidation: TicketValidating {
+			jwk = firstCBCKey
+		} else {
+			completion(.failure(.validationServiceDocument(.VS_ID_NO_ENC_KEY)))
+			return
+		}
 
-	}
-	
-	func requestServiceIdentityDocument(
-		validationServiceData: TicketValidationServiceData,
-		validationServiceJwkSet: [JSONWebKey],
-		completion: @escaping (Result<ServiceIdentityRequestResult, ServiceIdentityRequestError>) -> Void
-	) {
-		guard let url = URL(string: validationServiceData.serviceEndpoint) else {
-			completion(.failure(.UNKOWN))
+		/// 2. Determine `publicKeyForEncryption`
+
+		guard let publicKey = jwk.publicRSASecKey else {
+			completion(.failure(.other))
 			return
 		}
-		
-		let resource = ServiceIdentityDocumentResource(endpointUrl: url)
-		
-		restServiceProvider.load(resource) { [weak self] result in
-			switch result {
-			case .success(let serviceIdentityDocument):
-				self?.serviceIdentityProcessor.process(
-					validationServiceJwkSet: validationServiceJwkSet,
-					serviceIdentityDocument: serviceIdentityDocument,
-					completion: completion
-				)
-			case .failure(let error):
-				completion(.failure(.REST_SERVICE_ERROR(error)))
+
+		/// 3. Encrypt and sign DCC
+
+		guard let selectedHealthCertificate = selectedHealthCertificate, let accessTokenResult = accessTokenResult, let keyPair = keyPair, let validationDecoratorDocument = validationDecoratorDocument, let validationServiceDocument = validationServiceDocument else {
+			Log.error("validate called too early")
+			completion(.failure(.other))
+			return
+		}
+
+		let result = DCCEncryption().encryptAndSignDCC(
+			dccBarcodeData: selectedHealthCertificate.base45,
+			nonceBase64: accessTokenResult.nonceBase64,
+			encryptionScheme: encryptionScheme,
+			publicKeyForEncryption: publicKey,
+			privateKeyForSigning: keyPair.privateKey
+		)
+
+		switch result {
+		case .success(let encryptionResult):
+			self.encryptionResult = encryptionResult
+
+			requestResultToken(
+				serviceEndpoint: accessTokenResult.accessTokenPayload.aud,
+				validationServiceJwkSet: validationDecoratorDocument.validationServiceJwkSet,
+				validationServiceSignJwkSet: validationServiceDocument.validationServiceSignKeyJwkSet,
+				jwt: accessTokenResult.accessToken,
+				encryptionKeyKid: jwk.kid,
+				encryptedDCCBase64: encryptionResult.encryptedDCCBase64,
+				encryptionKeyBase64: encryptionResult.encryptionKeyBase64,
+				signatureBase64: encryptionResult.signatureBase64,
+				signatureAlgorithm: encryptionResult.signatureAlgorithm,
+				encryptionScheme: encryptionScheme
+			) { [weak self] resultTokenRequestResult in
+				switch resultTokenRequestResult {
+				case .success(let resultTokenResult):
+					self?.resultTokenResult = resultTokenResult
+					completion(.success(resultTokenResult.resultTokenPayload))
+				case .failure(let error):
+					completion(.failure(.resultToken(error)))
+				}
 			}
+		case .failure(let error):
+			completion(.failure(.encryption(error)))
 		}
 	}
 
+	func cancel() {
+
```

Comment:
We don't call server in this case?

Reply:
The story for graceful cancellation was postponed, so the implementation is empty for now